### PR TITLE
MOTECH-2052 Change BLOBs into join tables in motechschema db

### PR DESCRIPTION
**Pull request by martokarski**
I have problem with this tests:
EntityServiceContextIT.shouldSaveEntityWithGivenLookups
MdsBundleIT.testEntitiesBundleInstallsProperly

Tests fail during cleaining Lookup table from db:

javax.jdo.JDOUserException: Cannot write fields to a deleted object

FailedObject:47
	at org.datanucleus.api.jdo.state.PersistentDeleted.transitionWriteField(PersistentDeleted.java:125)
	at org.datanucleus.state.AbstractStateManager.transitionWriteField(AbstractStateManager.java:692)
	at org.datanucleus.state.StateManagerImpl.preWriteField(StateManagerImpl.java:4428)
	at org.datanucleus.state.StateManagerImpl.makeDirty(StateManagerImpl.java:2350)
	at org.datanucleus.store.types.wrappers.ArrayList.makeDirty(ArrayList.java:176)
	at org.datanucleus.store.types.wrappers.backed.ArrayList.clear(ArrayList.java:828)
	at org.datanucleus.store.rdbms.mapping.java.CollectionMapping.preDelete(CollectionMapping.java:296)
	at org.datanucleus.store.rdbms.request.DeleteRequest.execute(DeleteRequest.java:193)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.deleteTable(RDBMSPersistenceHandler.java:504)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.deleteObject(RDBMSPersistenceHandler.java:475)
	at org.datanucleus.state.AbstractStateManager.internalDeletePersistent(AbstractStateManager.java:832)
	at org.datanucleus.state.StateManagerImpl.deletePersistent(StateManagerImpl.java:4342)
	at org.datanucleus.ExecutionContextImpl.deleteObjectInternal(ExecutionContextImpl.java:2334)
	at org.datanucleus.store.rdbms.scostore.FKListStore.clear(FKListStore.java:754)
	at org.datanucleus.store.types.wrappers.backed.ArrayList.clear(ArrayList.java:839)
	at org.datanucleus.store.rdbms.mapping.java.CollectionMapping.preDelete(CollectionMapping.java:296)
	at org.datanucleus.store.rdbms.request.DeleteRequest.execute(DeleteRequest.java:193)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.deleteTable(RDBMSPersistenceHandler.java:504)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.deleteObject(RDBMSPersistenceHandler.java:475)
	at org.datanucleus.state.AbstractStateManager.internalDeletePersistent(AbstractStateManager.java:832)
	at org.datanucleus.state.StateManagerImpl.deletePersistent(StateManagerImpl.java:4342)
	at org.datanucleus.ExecutionContextImpl.deleteObjectInternal(ExecutionContextImpl.java:2334)
	at org.datanucleus.ExecutionContextImpl.deleteObjectWork(ExecutionContextImpl.java:2256)
	at org.datanucleus.ExecutionContextImpl.deleteObject(ExecutionContextImpl.java:2207)
	at org.datanucleus.api.jdo.JDOPersistenceManager.jdoDeletePersistent(JDOPersistenceManager.java:797)
	at org.datanucleus.api.jdo.JDOPersistenceManager.deletePersistent(JDOPersistenceManager.java:815)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.orm.jdo.TransactionAwarePersistenceManagerFactoryProxy$PersistenceManagerInvocationHandler.invoke(TransactionAwarePersistenceManagerFactoryProxy.java:211)
	at com.sun.proxy.$Proxy136.deletePersistent(Unknown Source)
	at org.motechproject.mds.repository.MotechDataRepository.delete(MotechDataRepository.java:163)
	at org.motechproject.mds.service.impl.EntityServiceImpl.commitChanges(EntityServiceImpl.java:430)
	at org.motechproject.mds.service.impl.EntityServiceImpl.commitChanges(EntityServiceImpl.java:603)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:318)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:110)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy135.commitChanges(Unknown Source)
	at org.motechproject.mds.it.osgi.MdsBundleIT.prepareTestEntities(MdsBundleIT.java:1092)
	at org.motechproject.mds.it.osgi.MdsBundleIT.testEntitiesBundleInstallsProperly(MdsBundleIT.java:230)
	at org.motechproject.mds.it.osgi.MdsBundleIT.testEntitiesBundleInstallsProperly(MdsBundleIT.java:213)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.ops4j.pax.exam.invoker.junit.internal.ContainerTestRunner.runChild(ContainerTestRunner.java:68)
	at org.ops4j.pax.exam.invoker.junit.internal.ContainerTestRunner.runChild(ContainerTestRunner.java:37)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:138)
	at org.ops4j.pax.exam.invoker.junit.internal.JUnitProbeInvoker.invokeViaJUnit(JUnitProbeInvoker.java:124)
	at org.ops4j.pax.exam.invoker.junit.internal.JUnitProbeInvoker.findAndInvoke(JUnitProbeInvoker.java:97)
	at org.ops4j.pax.exam.invoker.junit.internal.JUnitProbeInvoker.call(JUnitProbeInvoker.java:73)
	at org.ops4j.pax.exam.nat.internal.NativeTestContainer.call(NativeTestContainer.java:109)
	at org.motechproject.testing.osgi.container.MotechNativeTestContainer.call(MotechNativeTestContainer.java:106)
	at org.ops4j.pax.exam.spi.reactors.SingletonStagedReactor.invoke(SingletonStagedReactor.java:114)
	at org.ops4j.pax.exam.spi.reactors.PerSuiteStagedReactor.invoke(PerSuiteStagedReactor.java:47)
	at org.ops4j.pax.exam.junit.impl.ProbeRunner$2.evaluate(ProbeRunner.java:264)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:263)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:69)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:48)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:292)
	at org.ops4j.pax.exam.junit.impl.ProbeRunner.run(ProbeRunner.java:98)
	at org.ops4j.pax.exam.junit.PaxExam.run(PaxExam.java:93)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:24)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:292)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:24)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:231)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:60)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:50)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:222)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:292)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:157)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)

### CODE REVIEW

**Comment by kmadej** [comment]
The problem occurs when we are copying lookups from entity to draft. There is a copy() method in Lookup class. This method returns new Lookup object. Five fields (fieldsOrder, rangeLookupFields, setLookupFields, customOperators, useGenericParams) from Lookup are incorrectly passed to constructor. We should create new collection/map object with the same values. For example :

new ArrayList<>(getRangeLookupFields())

Then we will avoid this jdo exception while deleting lookup object from DB. 

Additionally, can you improve assert in shouldSaveEntityWithGivenLookups() test ? and remove redundant commitChanges invocation in MdsBundleIT.prepareTestEntities() ?

**Comment by pgesek** [comment]
test this psql

File: platform/mds/mds/src/main/resources/db/migration/mysql/V1__Base_version.sql [hunk 1]
```diff
@@ -49,6 +49,16 @@ CREATE TABLE "Entity_securityMembers" (
   CONSTRAINT "Entity_securityMembers_FK1" FOREIGN KEY ("Entity_OID") REFERENCES "Entity" ("id")
 );
 
+DROP TABLE IF EXISTS "EntityDraft_fieldNameChanges";
+CREATE TABLE "EntityDraft_fieldNameChanges" (
+  "id_OID" bigint(20) NOT NULL,
+  "key" varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
+  "value" varchar(255) CHARACTER SET utf8 COLLATE utf8_bin DEFAULT NULL,
+  PRIMARY KEY ("id_OID", "key"),
+  KEY "EntityDraft_fieldNameChanges_N49" ("id_OID"),
```
Comment (pgesek) [inline platform/mds/mds/src/main/resources/db/migration/mysql/V1__Base_version.sql]: can we come up with better names for these keys?

**Comment by martokarski** [comment]
test this psql

File: platform/mds/mds/src/main/resources/db/migration/mysql/V1__Base_version.sql [hunk 2]
```diff
@@ -49,6 +49,16 @@ CREATE TABLE "Entity_securityMembers" (
   CONSTRAINT "Entity_securityMembers_FK1" FOREIGN KEY ("Entity_OID") REFERENCES "Entity" ("id")
 );
 
+DROP TABLE IF EXISTS "EntityDraft_fieldNameChanges";
+CREATE TABLE "EntityDraft_fieldNameChanges" (
+  "id_OID" bigint(20) NOT NULL,
+  "oldName" varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
+  "newName" varchar(255) CHARACTER SET utf8 COLLATE utf8_bin DEFAULT NULL,
+  PRIMARY KEY ("id_OID", "oldName"),
+  KEY "EntityDraft_fieldNameChanges_N49" ("id_OID"),
```
Comment (pgesek) [inline platform/mds/mds/src/main/resources/db/migration/mysql/V1__Base_version.sql]: I meant these auto-generated keys ending with N49, can't we change their names to something better? For example EntityDraft_fieldNameChange_KeyIdx (as far as I understand this key basically means an index on the PK?)

**Pull request merged**